staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1012 @@
+				"has(embedded.spec)":               "undefined field 'spec'",
+			},
+		},
+		{name: "embedded object with properties",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XEmbeddedResource: true,
+					},
+					Properties: map[string]schema.Structural{
+						"kind":       stringType,
+						"apiVersion": stringType,
+						"spec": *objectTypePtr(map[string]schema.Structural{
+							"field1": stringType,
+						}),
+					},
+				},
+			}),
+			valid: []string{
+				"embedded.spec.field1 == 'a'",
+			},
+			// only field declared in the schema can be field selected in CEL expressions
+			errors: map[string]string{
+				"has(embedded.metadata)": "undefined field 'metadata'",
+			},
+		},
+		{name: "embedded with prune unknown",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Extensions: schema.Extensions{
+						XPreserveUnknownFields: true,
+						XEmbeddedResource:      true,
+					},
+				},
+			}),
+			errors: map[string]string{
+				// we only provide information in the schema to validation rules
+				// without type=object, we don't provide access to the embedded resource
+				"has(self.embedded)": "undefined field 'embedded'",
+			},
+		},
+		{name: "embedded object with prune unknown",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XPreserveUnknownFields: true,
+						XEmbeddedResource:      true,
+					},
+				},
+			}),
+			valid: []string{
+				"has(self.embedded)",
+			},
+			// only field declared in the schema can be field selected in CEL expressions
+			errors: map[string]string{
+				"has(embedded.metadata)": "undefined field 'metadata'",
```

Comment:
this should work

Reply:
`spec` below is correctly rejeted.

Reply:
Fixed in latest changes. `apiVersion`, `kind` and `metadata`are now always available. I added a couple test cases.